apps/src/componentLibrary/button/_baseButton/_BaseButton.tsx

```diff
@@ -0,0 +1,248 @@
+import classNames from 'classnames';
+import React, {memo, useMemo} from 'react';
+
+import {ButtonType, ButtonColor} from '@cdo/apps/componentLibrary/button';
+import FontAwesomeV6Icon, {
+  FontAwesomeV6IconProps,
+} from '@cdo/apps/componentLibrary/fontAwesomeV6Icon';
+import {ComponentSizeXSToL} from '@cdo/apps/componentLibrary/common/types';
+
+import moduleStyles from './_baseButton.module.scss';
+
+export interface _BaseButtonProps {
```

Comment:
I was thinking a little more about how to organize these props and enforce some of these contracts through typescript...here's an idea if you'd like to do something similar:

```
// First, create the common props and extensions for each Button variant

interface CommonButtonProps {
  className?: string;
  id: string;
  disabled?: boolean;
  isPending?: boolean;
  // ... others
}

interface LinkButtonProps extends CommonButtonProps {
  href: string;
  download?: string | boolean;
  // ... others
}

interface ButtonProps extends CommonButtonProps {
  onClick: () => void;
  buttonType?: 'submit' | 'button';
  // ... others
}

// Then, we can create separate interfaces to distinguish between buttons with text and buttons with an icon only

interface IconProps {
  icon: FontAwesomeV6IconProps;
}

interface TextProps {
  text: string;
  iconLeft?: FontAwesomeV6IconProps;
  iconRight?: FontAwesomeV6IconProps;
}

// We can use this XOR utility (found on SO, link below) to make sure that  the combined props use text or an icon but not both

type Without<T, U> = {[P in Exclude<keyof T, keyof U>]?: never};
type XOR<T, U> = T | U extends object
  ? (Without<T, U> & U) | (Without<U, T> & T)
  : T | U;

type CombinedButtonProps<T> = XOR<T & IconProps, T & TextProps>;

// Finally, create our Button variants

const Button: React.FunctionComponent<CombinedButtonProps<ButtonProps>> = props => { ... };

const LinkButton: React.FunctionComponent<CombinedButtonProps<LinkButtonProps>> = props => { ... };


// Now TS should be able to check at compile time that we're using the right prop combinations:


// X Doesn't compile - missing text and/or icon
<Button id="id" onClick={() => {}} />;

// ✓ Compiles - using only icon
<Button id="id" onClick={() => {}} icon={{iconName: 'icon'}} />;

// ✓ Compiles - using only text
<Button id="id" onClick={() => {}} text="text" />;

// X Doesn't compile - using both text and icon
<Button id="id" onClick={() => {}} icon={{iconName: 'icon'}} text="text" />;

// X Doesn't compile - using iconLeft (text only prop) and icon
<Button id="id" onClick={() => {}} icon={{iconName: 'icon'}} iconLeft={{iconName: 'icon'}} />;


// X Doesn't compile - missing text and/or icon
<LinkButton id="id" href="href" />;

// ✓ Compiles - using only icon
<LinkButton id="id" href="href" icon={{iconName: 'icon'}} />;

// ✓ Compiles - using only text
<LinkButton id="id" href="href" text="text" />;

// X Doesn't compile - using both text and icon
<LinkButton id="id" href="href" icon={{iconName: 'icon'}} text="text" />;

// X Doesn't compile - using iconRight (text only prop) and icon
<LinkButton id="id" href="href" icon={{iconName: 'icon'}} iconRight={{iconName: 'icon'}} />;
```

XOR type found on SO: https://stackoverflow.com/questions/42123407/does-typescript-support-mutually-exclusive-types

Reply:
Great idea and structure, thank you!

Though, not sure that will help, since I've tried XOR logic before and that lead to types collisions in BaseButton still.
Might work for Button and LinkButton, but will still cause typerrors with _BaseButton probably

Reply:
<img width="1720" alt="image" src="https://github.com/code-dot-org/code-dot-org/assets/22244040/cb3d6a7a-897e-49f1-9dda-9b7cdc798d27">
Checked locally, unfortunately it didn't help and caused type errors in `_BaseButton` component again.

I've reorganized the structure of types to be reusable and better documented, I think it might be better to merge as is